modules/apps/frontend-data-set/frontend-data-set-impl/src/main/resources/com/liferay/frontend/data/set/internal/batch/01-object-definition.batch-engine-data.json

```diff
@@ -232,6 +232,24 @@
 				}
 			],
 			"objectRelationships": [
+				{
+					"deletionType": "cascade",
+					"edge": true,
+					"externalReferenceCode": "L_DATA_SET_TO_DATA_SET_ACTIONS",
+					"label": {
+						"en_US": "Data Set to Data Set Actions"
+					},
+					"name": "dataSetToDataSetActions",
+					"objectDefinitionExternalReferenceCode1": "L_DATA_SET",
+					"objectDefinitionExternalReferenceCode2": "L_DATA_SET_ACTION",
+					"objectDefinitionName2": "DataSetAction",
+					"objectDefinitionSystem2": true,
+					"parameterObjectFieldId": 0,
+					"parameterObjectFieldName": "",
+					"reverse": false,
+					"system": true,
+					"type": "oneToMany"
+				},
```

Comment:
> Adding a new dummy relationship between Data Set and Data Set Action entities. Rationale is that root model requires relations of type edge to be created, which are mandatory when connecting object entries. We use 2 existing connections to model "creation" and "item" actions, that are not mandatory. As we want to keep them like that, we need an extra connection, not used to fetch data.

This seems to me very hacky, the type of solution that got out earlier 2-step definition blocked by Brian. Where is the inadequacy coming from? Is the problem in root model implementation or `edge` implementation in batch engine? Why would `edge` setting result in anything being mandatory? Why are relationships causing anything to be mandatory? Attributes/fields of a database table/object definition are the ones that are required or not. But to my understanding, relationships are never required or not, they just are. Their inclusion or omission is something that is handled in business logic, on app level.

Reply:
A solution to consider, if we get stuck on this, is to drop two relationships, and query actions filtered by `type`.

Reply:
This looks an internal constraint in the root model implementation. When one "binds" object definitions, a relation has to be specified. That one becomes "edge" as per my testing in the UI + exporting definitions.

So to make this work in batch, the `edge` must be set to true in the JSON

Once that happens, creating an object entry for the "child" object definition (`Data Set Action` in our case) implies connecting it right away with some "parent" entry (`Data Set` in our case) via the `edge` relationship, so such connections can't be optional.

It makes "sense" because a non root entity should not exist in isolation.

Main oversight on the object root model implementation, imho, is that when there are more than 1 connection between the same pair of definitions, only one of it is used to create the root model. That makes us to choose which one of the connection is made part of the root. 

This is not wrong per-se, instead, problem is that all `edge` connections must be supplied. so if we have
```
    /------------- connection 1 ------------\
A →|                                         |→ B
    \------------- connection 2 ------------/
```
Being `A` and `B` two object definitions, and we use `connection 1` to create the root between `A` and `B`, then `connection 1` is marked as `edge`, and thus becomes mandatory when you connect an instance of `A` with an instance of `B`. We don't want that in our data model.

We can make both `connection 1` and `connection 2` as `edges`, but then, instead of making just any of them mandatory for every pair of entries, both of them become mandatory for any pair of entries. In our model, this means if we create a `Data Set Action` we'll forced to connect it to `Data Set` both with `Data Set Creation Action` and `Data Set Item Action` connections.

Here, root model implementation needs to be more flexible because we will connect an action with a data set using always one of such `edge` connections, but not both at the same time.

Reply:
> We can make both connection 1 and connection 2 as edges, but then, instead of making just any of them mandatory for every pair of entries, both of them become mandatory for any pair of entries. In our model, this means if we create a Data Set Action we'll forced to connect it to Data Set both with Data Set Creation Action and Data Set Item Action connections at the same time.

This is the part that I don't understand, why is it mandatory to connect to both? All the relationships configuration exists in the backend / persistence, so permissions for an action on leaf should be checkable on root. Why does root model implementation care if a relationship is omitted from request? It should be able to check any relevant relationships for root model, and return partial child data from given relationship in request.

I think we should communicate this to the BPM team, even if we make a workaround here. A customer or partner will run into this sooner or later, so the team should be aware of it.

As for the workaround / alternative solution, I agree that the attribute approach seems better (the last one in chart above). Multiple entities make more sense if there are large differences in attributes between them, like it is for `*Filter` entities. In this case, all three are identical, as evidenced by the fact all could have the underlying `DataSetAction` entity.

I wouldn't mind `role` for the attribute name. It fits purpose from DSM perspective. It's not ideal because `Creation` and `Item/Bulk` do not describe the action on the same axis. `Creation` describes what the action _does_,  while `Item` describes what the action _affects_, what's the scope of its use. To illustrate naming on the same level, we could have 
- `Creation`, `Edit` and `Deletion` actions - or-
- `Collection`, `Item` and `Bulk / Multiple Items` actions.

Since creation action is the only collection action we are using in DSM, it does make sense to my simplify it in UI. But the effect on code naming is a bit awkward, as we cannot cleanly map code to use, unless we use generic terms like `type` and `role`.

All in all, I'd say an alternative to `role` would be something like `scope`, with values `collection`, `item` and `multipleItems`. But again, a generic `role` with a clearer connection to product UI, seems good as well.

Reply:
Yep, semantically, we can look at the `role` property from various perspectives. I like to see them from the pov of the thing being targeted
 - Creation actions exist at a prior level as it does not target an instance/entry but rather can be said it targets the "type" of the resource being created (e.g. the object definition). This is how we modeled permissions, and explains why it's complicated to use a name for the thing being _affected_ by the action. It does not affect a "collection" either because when there are no items, there is no collection. When one "creates an entry" (s)he is doing so for a given "type" of thing. 
 - Item actions target exactly one instance of the resource
 - Bulk actions target one or more instances of the resource, determined by the current selection

The problem with the term _target_ is that it's already taken in the FDS, and to make things worse, DSM uses _type_ to [denote the very same thing](https://github.com/liferay/liferay-portal/blob/a3958a054163b220e064e5c3e89d91316a80e3a0/modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java#L389)

So we're left with fewer options here. We can consider options like:

- _scope_ , focusing on action target, with values `resource type`, `single instance` and `instance selection`. This is perhaps the most formal
- _role_, focused on a generic notion of "action type", with values `creation`, `item`, `collection` 
- _group_ / _part-of_, focused on the physical placement of the action, with values `CreationMenu`, `ItemDropdown`, `BulkMenu`

None of these will keep the actual mapping because FDS API is not so consistent.

Reply:
My vote is also for the third option, the attribute one. Seems to be the less _hacky_ and maybe more future proof if there is a need of adding more values without breaking the model.

Regarding the naming, nothing other than possible synonyms of type like `category`, `family` or `suite` come to mind.

Reply:
I think I like `category` here the most. `role` may be misleading into thinking value has something to do with permissions. 

Alternatively, we can add `target`, switch current `type` to work with `target`, and use `type` for the new purpose. The reason I initially set `type` instead of `target` is because `target` in FDS is misleading into thinking it has anything to do with `event.target`. But, if we change to use in DSM, at least our APIs will be consistent.

Reply:
This change from `type` to `target` sounds good 👌

Reply:
thanks guys, I'm leaning towards freeing the type attribute:

- Rename current `type` to `target`. This aligns the APIs
- Use `type` to encode the type of action

This kind of breaking changes would not be possible after beta